src/dune_engine/build_system.ml

```diff
@@ -539,10 +539,12 @@ end = struct
               let remove_target_dir dir = Path.rm_rf (Path.build dir) in
               let remove_target_file path =
                 try Path.Build.unlink path with
-                | Unix.Unix_error (EISDIR, _, _) ->
-                  (* If target changed from a directory to a file, delete
-                     in anyway. *)
-                  remove_target_dir path
+                (* If target changed from a directory to a file, delete
+                   in anyway. *)
+                | Unix.Unix_error (error, _, _) ->
+                  (match error, Platform.OS.value with
+                   | EISDIR, _ | EPERM, Darwin -> remove_target_dir path
```

Comment:
This is because `unlink` actually errors with `EPERM` on macos when it's passed a directory, not `EISDIR`. Shall I abstract that to a new function that returns a variant for this case?

Reply:
I think this deserves a comment.

Reply:
Actually, never mind: now that it's a separate function, it seems clear enough.